platform/mbed_error.h

```diff
@@ -23,13 +22,782 @@
+    THREAD_DELETE_FAILED       302      Thread Delete failed 
+    PROHIBITED_IN_ISR_CONTEXT  303      Operation Prohibited in ISR context 
+    PINMAP_INVALID             304      Pinmap Invalid 
+    RTOS_EVENT                 305      Unknown Rtos Error 
+    RTOS_THREAD_EVENT          306      Rtos Thread Error 
+    RTOS_MUTEX_EVENT           307      Rtos Mutex Error 
+    RTOS_SEMAPHORE_EVENT       308      Rtos Semaphore Error 
+    RTOS_MEMORY_POOL_EVENT     309      Rtos Memory Pool Error 
+    RTOS_TIMER_EVENT           310      Rtos Timer Error 
+    RTOS_EVENT_FLAGS_EVENT     311      Rtos Event flags Error 
+    RTOS_MESSAGE_QUEUE_EVENT   312      Rtos Message queue Error 
+    DEVICE_BUSY                313      Device Busy 
+    CONFIG_UNSUPPORTED         314      Configuration not supported 
+    CONFIG_MISMATCH            315      Configuration mismatch 
+    ALREADY_INITIALIZED        316      Already initialzied 
+    HARDFAULT_EXCEPTION        317      HardFault exception 
+    MEMMANAGE_EXCEPTION        318      MemManage exception 
+    BUSFAULT_EXCEPTION         319      BusFault exception 
+    USAGEFAULT_EXCEPTION       320      UsageFault exception
+    \endverbatim
+ *
+ *  @note
+ *  Custom Error codes can be defined using the macro DEFINE_CUSTOM_ERROR\n
+ *  This is mainly meant to capture non-generic error codes specific to a device.
+ *  For example DEFINE_CUSTOM_ERROR( MY_CUSTOM_ERROR ,1 ) macro defines the following values:\n
+ *      ERROR_CODE_MY_CUSTOM_ERROR = MBED_CUSTOM_ERROR_BASE+1\n
+ *      ERROR_MY_CUSTOM_ERROR = MAKE_MBED_ERROR(ERROR_TYPE_CUSTOM, ENTITY_UNKNOWN, ERROR_CODE_MY_CUSTOM_ERROR)\n
+ *  Its effectively equivalent to:\n
+ *      ERROR_CODE_MY_CUSTOM_ERROR = 4097\n
+ *      ERROR_MY_CUSTOM_ERROR = 0xA0FF1001\n (Note that ENTITY field is set to ENTITY_UNKNOWN) \n\n
+ *
+ *  
+ *  @note
+ *  **Searching for error codes in mbed-os source tree:** \n
+ *  If you get an error report as below which you want to search for in mbed-os source tree, first take note of "Error Code" number. \n
+ *  For example, the below error report has an error code of \b 259. Find the error name associated with the error code and in this case its \b INVALID_FORMAT. \n
+ *  Use that error name(\b INVALID_FORMAT) to search the source tree for code locations setting that specific error code. \n 
+ *  If the Error Entity reported is not 255(which indicates unknown entity), you can also use that to narrow down to the specific component reporting the error.
+ *  See MbedEntityType enum above for entity mapping. \n
+ *  
+ *  \verbatim
+    ++ MbedOS Error Info ++
+    Error Status: 0x80040103
+    Error Code: 259
+    Error Entity: 04
+    Error Message: HAL Entity error
+    Error Location: 0x000067C7
+    Error Value: 0x00005566
+    Current Thread: Id: 0x200024A8 EntryFn: 0x0000FB0D StackSize: 0x00001000 StackMem: 0x200014A8 SP: 0x2002FFD8
+    -- MbedOS Error Info --
+    \endverbatim
+ */
+ 
+typedef enum _MbedErrorCode
```

Comment:
That's intimidating, we should also set some guidelines inside the system which errors should be used. POSIX errors for POSIX functions and system errors for the rest seems reasonable.

Reply:
Ok, thats good suggestion, Ill document that here and in handbook.